metrics/server.go

```diff
@@ -0,0 +1,384 @@
+		http.Error(w, fmt.Sprintf("unknown application %s", application), http.StatusBadRequest)
+		return
+	}
+	log.Logger.Tracef("getMetrics found routemap %v", rm)
+
+	// initialize result
+	result := make(map[string]*MetricSummary, 0)
+	byMetricOverTransactions := make(map[string](map[string][]float64), 0)
+	byMetricOverUsers := make(map[string](map[string][]float64), 0)
+
+	// for each version:
+	//   get metrics
+	//   for each metric, compute summary for metric, version
+	//   prepare for histogram computation
+	for v, version := range rm.GetVersions() {
+		signature := version.GetSignature()
+		if signature == nil {
+			log.Logger.Debugf("no signature for application %s (version %d)", application, v)
+			continue
+		}
+
+		versionmetrics, err := abn.MetricsClient.GetMetrics(application, v, *signature)
+		if err != nil {
+			log.Logger.Debugf("no metrics found for application %s (version %d; signature %s)", application, v, *signature)
+			continue
+		}
+
+		for metric, metrics := range *versionmetrics {
+			_, ok := result[metric]
+			if !ok {
+				// no entry for metric result; create empty entry
+				result[metric] = &MetricSummary{
+					HistogramsOverTransactions: nil,
+					HistogramsOverUsers:        nil,
+					SummaryOverTransactions:    []*VersionSummarizedMetric{},
+					SummaryOverUsers:           []*VersionSummarizedMetric{},
+				}
+			}
+
+			entry := result[metric]
+
+			smT, err := calculateSummarizedMetric(metrics.MetricsOverTransactions)
+			if err != nil {
+				log.Logger.Debugf("unable to compute summaried metrics over transactions for application %s (version %d; signature %s)", application, v, *signature)
+				continue
+			} else {
+				entry.SummaryOverTransactions = append(entry.SummaryOverTransactions, &VersionSummarizedMetric{
+					Version:          v,
+					SummarizedMetric: smT,
+				})
+			}
+
+			smU, err := calculateSummarizedMetric(metrics.MetricsOverUsers)
+			if err != nil {
+				log.Logger.Debugf("unable to compute summaried metrics over users for application %s (version %d; signature %s)", application, v, *signature)
+				continue
+			}
+			entry.SummaryOverUsers = append(entry.SummaryOverUsers, &VersionSummarizedMetric{
+				Version:          v,
+				SummarizedMetric: smU,
+			})
+			result[metric] = entry
+
+			// copy data into structure for histogram calculation (to be done later)
+			// over transaction data
+			vStr := fmt.Sprintf("%d", v)
+			_, ok = byMetricOverTransactions[metric]
+			if !ok {
+				byMetricOverTransactions[metric] = make(map[string][]float64, 0)
+			}
+			(byMetricOverTransactions[metric])[vStr] = metrics.MetricsOverTransactions
+
+			_, ok = byMetricOverUsers[metric]
+			if !ok {
+				byMetricOverUsers[metric] = make(map[string][]float64, 0)
+			}
+			(byMetricOverUsers[metric])[vStr] = metrics.MetricsOverUsers
+		}
+	}
+
+	// compute histograms
+	for metric, byVersion := range byMetricOverTransactions {
+		hT, err := calculateHistogram(byVersion, 0, 0)
```

Comment:
Are we going to use `bucketSize` 10 and `decimalPlace` 1?

Reply:
good catch. fixing.